CIP-0057/README.md

```diff
@@ -0,0 +1,414 @@
+
+##### fields
+
+This keyword's value must be an array of valid _Plutus Data Schema_; possibly empty. It must also always be specified with the `index` keyword. An instance is valid against this keyword if it represents a Plutus constructor for which each field is valid under each subschema given by this keyword's value. Fields are compare positionally. 
+
+## Example(s)
+
+<details>
+  <summary>Lobster Challenge Contract</summary>
+
+```json
+{
+  "$schema": "https://json-schema.org/draft/2020-12/schema",
+
+  "$id": "https://cips.cardano.org/cips/cip57/schemas/v1/lobster-challenge.json",
+
+  "$vocabulary": {
+    "https://json-schema.org/draft/2020-12/vocab/core": true,
+    "https://json-schema.org/draft/2020-12/vocab/applicator": true,
+    "https://json-schema.org/draft/2020-12/vocab/validation": true,
+    "https://cips.cardano.org/cips/cip57": true
+  },
+
+  "preamble": {
+    "title": "Lobster Challenge",
+    "description": "A specification of the lobster challenge contract: <https://github.com/input-output-hk/lobster-challenge/>",
+    "version": "1.0.0",
+    "license": "Apache-2.0"
+  },
+  "validators": {
+    "main": {
+      "description": "The lobster contract validator, verifying vote counts and driving token evolution.",
+      "datum": {
+        { "$ref": "plutus-types.json#/$defs/Unit" }
+      },
+      "redeemer": {
+        { "$ref": "plutus-types.json#/$defs/Unit" }
+      },
+      "compiledCode": {
+        "template": "59094c01000{{ lobsterParams }}006002003",
+        "arguments": {
+          "lobsterParams": {
+            "$ref": "#/definitions/LobsterParams" 
+          }
+        }
+      }
+    },
+    "contractIdPolicy": {
+      "description": "A minting policy to create a unique thread token for the contract.",
+      "compiledCode": {
+        "template": "59094c01000{{ tokenName }}{{ txOutRef }}00402",
+        "arguments": {
+          "tokenName": { "$ref": "plutus-types.json#/$defs/AssetName" },
+          "txOutRef": { "$ref": "plutus-types.json#/$defs/TransactionOutputReference" }
+        }
+      }
+    },
+    "alwaysTruePolicy": {
+      "description": "A minting policy that is always true.",
+      "compiledCode": {
+        "template": "59094c01000006002003"
+      }
+    }
+  },
+  "endpoints": {
+    "init": {
+      "description": "Initialize the contract instance. The state is carried not by datums, but by token values minting with the 'alwaysTruePolicy'", 
+      "from": null,
+      "to": { "$refs": "#/validators/main/datum" },
+      "via": { "$refs": "#/validators/main/redeemer" },
+    },
+    "vote": {
+      "description": "A vote that is increasing the voting count and the name counter by minting more tokens.", 
+      "from": { "$refs": "#/validators/main/datum" },
+      "to": { "$refs": "#/validators/main/datum" },
+      "via": { "$refs": "#/validators/main/redeemer" },
+    }
+  },
```

Comment:
Are these endpoints meant to be implemented like that in the actual off-chain code? To me it seems like these information are not sufficient. A validator could also depend on the ScriptContext e.g. valid time range. Or it is required to spend a certain input to prove ownership of token X, etc.   
But maybe I understand this wrong here, happy to be corrected.

Reply:
What do you mean by "implemented like that" ? 

This is mainly meant to be a description of the various contract endpoints, and ideally yes, it should contain enough information to allow to fully construct a transaction from the specification (and user inputs). It is insufficient at the moment indeed, which I touch upon in the rationale. There are additional constraints of the transaction that needs to be expressed somehow. I think that the tx-constraints from plutus-tx offers a quite nice abstraction for this and it seems possible to me to translate that in plain JSON. 

Ideally, I would like this effort to be driven by one or more real-life use case, trying to fully specify an existing contract and address see where the shortcomings are in the current specification.

Reply:
> What do you mean by "implemented like that" ?

Do these endpoints represent the unit state machine transition or endpoints which are practically used. The reason why I'm saying this is because an endpoint in the actual off-chain code could have multiple transitions and also be conditional.

Reply:
They represent how an application should interact with the contract; so they describe much more the off-chain interactions and conditions, as well as additional constraints that apply. 

Conceptually, a long-term could be to have UI automatically generated from a blueprint, all users would have to do is connect a wallet.